prow/config/cache.go

```diff
@@ -0,0 +1,388 @@
+	if err != nil {
+		return nil, err
+	}
+
+	if err := DefaultAndValidateProwYAML(c, prowYAML, identifier); err != nil {
+		return nil, err
+	}
+
+	return append(c.GetPostsubmitsStatic(identifier), prowYAML.Postsubmits...), nil
+}
+
+// GetProwYAMLFromCache uses ProwYAMLCache to first try to perform a lookup of
+// previously-calculated ProwYAML objects. The 'valConstructor' function is
+// taken as an argument to make it easier to test this function. This way, unit
+// tests can just provide its own function for constructing a ProwYAML object
+// (instead of needing to create an actual Git repo, etc.).
+func GetProwYAMLFromCache(
+	prowYAMLCache *ProwYAMLCache,
+	valConstructorHelper func(git.ClientFactory, string, RefGetter, ...RefGetter) (*ProwYAML, error),
+	gc git.ClientFactory,
+	identifier string,
+	baseSHAGetter RefGetter,
+	headSHAGetters ...RefGetter) (*ProwYAML, error) {
+
+	keyConstructor := func() (CacheKey, error) {
+		kp, err := MakeCacheKeyParts(identifier, baseSHAGetter, headSHAGetters...)
+		if err != nil {
+			return CacheKey(""), err
+		}
+
+		return MakeCacheKey(kp)
+	}
+
+	// The point of valConstructor is to allow us to mock this expensive value
+	// constructor call in tests (and, e.g., avoid having to do things like
+	// going over the network or dealing with an actual Git client, etc.).
+	valConstructor := func() (interface{}, error) {
+		return valConstructorHelper(gc, identifier, baseSHAGetter, headSHAGetters...)
+	}
+
+	val, err := prowYAMLCache.GetOrAdd(keyConstructor, valConstructor)
+	if err != nil {
+		return nil, err
+	}
+
+	prowYAML, ok := val.(*ProwYAML)
+	if ok {
+		return prowYAML, err
+	}
+
+	// Somehow, the value retrieved with GetFromCache has a malformed type. This
+	// can happen if some other function modified the cache. Ultimately, this is
+	// a price we pay for using a cache library that uses "interface{}" for the
+	// type of its items. In this case, we log a warning and return an error.
+	err = fmt.Errorf("cache value type error: expected value type '*config.ProwYAML', got '%T'", val)
+	logrus.Warn(err)
+	return nil, err
+}
+
+// GetOrAdd tries to use a cache if it is available to get a Value. It is
+// assumed that Value is expensive to construct from scratch, which is the
+// reason why we try to use the cache in the first place. If we do end up
+// constructing a Value from scratch, then we store it into the cache with a
+// corresponding Key, so that we can look up the Value with just the Key in the
+// future.
+//
+// This cache is resistant to cache stampedes because it uses a duplicate
+// suppression strategy. This is also called request coalescing.
+func (lruCache *LRUCache) GetOrAdd(
+	keyConstructor keyConstructor,
+	valConstructor valConstructor) (interface{}, error) {
+
+	// If the cache is unreachable, then fall back to cache-less behavior
+	// (construct the value from scratch).
+	if lruCache == nil {
```

Comment:
nit: I'm not sure it makes sense to support using the `Get*FromCache` functions at all if the cache hasn't been initialized. Rather than optionally initializing a field and then checking if it initialized, it might be cleaner to handle and use the cache directly so that if the calls are made the cache must exist.
Not that important either way though, just something to consider.

Reply:
I agree with your concern, because I don't think it's unreasonable to expect that users of `Get*FromCache` have already taken care of setting up the cache. I did some refactoring to move this nilcheck to downstream users (into `GetPresubmitsMaybeCached` and `GetPostsubmitsMaybeCached`).